pallets/identity/src/lib.rs

```diff
@@ -439,14 +439,17 @@ decl_module! {
         /// Leave the secondary key's identity.
         #[weight = <T as Config>::WeightInfo::leave_identity_as_key()]
         pub fn leave_identity_as_key(origin) -> DispatchResult {
-            let data = Self::ensure_origin_call_permissions(origin)?;
-            Self::leave_identity(Signatory::Account(data.sender), data.primary_did)
+            let (sender, did) = Self::ensure_did(origin)?;
+            Self::leave_identity(Signatory::Account(sender), did)
         }
 
         /// Leave an identity as a secondary identity.
         #[weight = <T as Config>::WeightInfo::leave_identity_as_identity()]
         pub fn leave_identity_as_identity(origin, did: IdentityId) -> DispatchResult {
-            let sender_did = Self::ensure_perms(origin)?;
+            let (sender, sender_did) = Self::ensure_did(origin)?;
+            if sender != DidRecords::<T>::get(sender_did).primary_key {
+                CallPermissions::<T>::ensure_call_permissions(&sender)?;
+            }
```

Comment:
This change is still not needed.  The new tests pass with the original permissions check:
`let sender_did = Self::ensure_perms(origin)?;`